From b5488eb15aca055c42122f140dd34f213f7b0e02 Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Wed, 22 May 2019 17:01:09 +0200
Subject: [PATCH 255/386] net: phy: phylink: propagate only link up event to
 SFP

The link down event must be triggered as until
recently, i.e. from the SFP layer. Previous patches caused issue,
when flush_work() was run from the interrupt context, which is
forbidden in Linux.

Execute the SFP update only for the link-up event and improve the
RNTL locking.

Change-Id: I2e1c2c08e8686073c2595b878b0b21b39eead391
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/9825
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Stefan Chulski <Stefan.Chulski@cavium.com>
[RH: Original patch taken from marvell 88F3720 board support SDK 10.0-PR2003]
Signed-off-by: Ruiqiang Hao <Ruiqiang.Hao@windriver.com>
---
 drivers/net/phy/phylink.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/net/phy/phylink.c b/drivers/net/phy/phylink.c
index 7a216b0f5a1f..4a33eedb8027 100644
--- a/drivers/net/phy/phylink.c
+++ b/drivers/net/phy/phylink.c
@@ -904,11 +904,14 @@ void phylink_mac_change(struct phylink *pl, bool up)
 
 #if IS_ENABLED(CONFIG_SFP)
 	/* Make sure the event is propagated to the SFP layer. */
-	if (pl->sfp_bus && rtnl_is_locked()) {
-		if (up)
+	if (pl->sfp_bus && up) {
+		if (rtnl_is_locked()) {
 			sfp_link_up(pl->sfp_bus);
-		else
-			sfp_link_down(pl->sfp_bus);
+		} else {
+			rtnl_lock();
+			sfp_link_up(pl->sfp_bus);
+			rtnl_unlock();
+		}
 	}
 #endif
 
-- 
2.17.1

